processing/src/main/java/org/apache/druid/query/context/ResponseContext.java

```diff
@@ -236,95 +253,124 @@ public static ResponseContext createEmpty()
-   * The resulting string might be correctly deserialized as a {@link ResponseContext}.
+   * This method tries to remove some elements from context collections if it's needed to satisfy the limit.
+   * The resulting string might be correctly deserialized to {@link ResponseContext}.
    */
   public SerializationResult serializeWith(ObjectMapper objectMapper, int maxCharsNumber) throws JsonProcessingException
   {
     final String fullSerializedString = objectMapper.writeValueAsString(getDelegate());
     if (fullSerializedString.length() <= maxCharsNumber) {
       return new SerializationResult(fullSerializedString, fullSerializedString);
     } else {
-      final HashMap<String, Object> copiedMap = new HashMap<>(getDelegate());
-      final PriorityQueue<Map.Entry<String, String>> serializedValueEntries = new PriorityQueue<>(
-          Comparator.comparing((Map.Entry<String, String> e) -> e.getValue().length()).reversed()
-      );
-      for (Map.Entry<String, Object> e : copiedMap.entrySet()) {
-        serializedValueEntries.add(new AbstractMap.SimpleImmutableEntry<>(
-            e.getKey(),
-            objectMapper.writeValueAsString(e.getValue())
-        ));
-      }
-      // quadratic complexity: while loop with map serialization on each iteration
-      while (!copiedMap.isEmpty() && !serializedValueEntries.isEmpty()) {
-        final Map.Entry<String, String> maxLengthEntry = serializedValueEntries.poll();
-        Preconditions.checkNotNull(maxLengthEntry);
-        copiedMap.remove(maxLengthEntry.getKey());
-        final String reducedSerializedString = objectMapper.writeValueAsString(copiedMap);
-        if (reducedSerializedString.length() <= maxCharsNumber) {
-          return new SerializationResult(reducedSerializedString, fullSerializedString);
+      // Indicates that the context is truncated during serialization.
+      add(Key.TRUNCATED, true);
+      final ObjectNode contextJsonNode = objectMapper.valueToTree(getDelegate());
+      final ArrayList<Map.Entry<String, JsonNode>> sortedNodesByLength = Lists.newArrayList(contextJsonNode.fields());
+      final Comparator<Map.Entry<String, JsonNode>> valueLengthReversedComparator =
+          Comparator.comparing((Map.Entry<String, JsonNode> e) -> e.getValue().toString().length()).reversed();
+      sortedNodesByLength.sort(valueLengthReversedComparator);
+      int needToRemoveCharsNumber = fullSerializedString.length() - maxCharsNumber;
+      // The complexity of this block is O(n*m*log(m)) where n - context size, m - context's array size
+      for (Map.Entry<String, JsonNode> e : sortedNodesByLength) {
+        final String fieldName = e.getKey();
+        final JsonNode node = e.getValue();
+        if (node.isArray()) {
+          if (needToRemoveCharsNumber >= node.toString().length()) {
+            final int lengthBeforeRemove = node.toString().length();
+            // Empty array could be correctly deserialized so we remove only its elements.
```

Comment:
I think the logic of this block should avoid producing empty array because it may be misleading.

Reply:
Removed empty arrays